### PR TITLE
Switch to Differential's `ord_neu` for key spines.

### PR DESCRIPTION
**Pull request by antiguru**
Switch over to Differential's `ord_neu` implementation also for key-spines. The PR reverts the `HeapSize` trait and implementations because we don't need it anymore, and then continues to clean up a few things so that it aligns well with the upstream changes.

### Motivation

Part of #22678.

### Tips for reviewer

Review commit-by-commit.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: test/testdrive/introspection-sources.td
Comment (vmarcos) [inline test/testdrive/introspection-sources.td:374]: It is odd to change the middle estimate here without changing the argument in the comment above (lines 333-342). Do we really need to change the middle estimate or just enlarge the range? If the former, would you provide a justification in the comment as well?

Also, @def- : Would you like to take this test for another spin before we merge this PR to avoid a regression of #23509?

**Review by vmarcos — COMMENTED** [review]
The changes to pick up `ord_neu` look fine to me, though I only partly looked into the new code in DD.

I was a bit unsure about a full revert of the `HeapSize` changes, as the interface and some of the implementations could be useful for attribution elsewhere. For example, #23202 would in principle not require moving to region-allocated structures; we could use the same interface and support for `Row` there.

I can feel the annoyance of not being able to provide a blanket implementation for `T: Copy`, though. Do you think that we should design another mechanic for non-region allocated attributions in lieu of `HeapSize` or would it be manageable to just specialize its implementation to the offset type we need?

File: test/testdrive/introspection-sources.td
Comment (antiguru) [reply test/testdrive/introspection-sources.td:374]: I'm more and more convinced that we should not try to have precise measurements here. The sizes reported here are eventually-consistent and highly depend on other parts of the code. Changes in Differential often change size, capacity, allocations slightly, and it's hard to connect the dots. For this reason, I'd like to move the tests to an order-of-magnitude approach where the values should make sense, but not precise.

**Comment by vmarcos** [comment]
> > I was a bit unsure about a full revert of the HeapSize changes, as the interface and some of the implementations could be useful for attribution elsewhere. For example, #23202 would in principle not require moving to region-allocated structures; we could use the same interface and support for Row there.
> 
> I agree! The reason why I removed it is because after switching to `ord_neu` it would not be used anymore, and I prefer to have no unused code in the repo and bring it back later when needed.

Yes, we could bring it back later, but why not use `HeapSize` for the offset `Vec`s? That would increase consistency with following the trait for estimating size of non-region allocated data structures and otherwise relying on the precise measurements from region-allocated structures.

**Review by vmarcos — APPROVED** [review]
As we agreed on the `HeapSize` revert, approved pending the fixes to `introspection-sources.td` discussed [here](https://github.com/MaterializeInc/materialize/pull/23612#discussion_r1414027918).

**Comment by antiguru** [comment]
Thanks for the reviews!

**Pull request merged**
